include/graphblas/base/internalops.hpp

```diff
@@ -39,9 +39,127 @@ namespace grb {
 
 	namespace operators {
 
+
+
 		/** Core implementations of the standard operators in #grb::operators. */
 		namespace internal {
 
+			/**
+			 * Standard negation operator.
+			 *
+			 * Assumes native availability of ! on the given data types or assumes that
+			 * the relevant operators are properly overloaded.
+			 *
+			 * @tparam Op The Operator class to negate.
+			 * 			Requires the following typedefs:
+			 * 			- \b D1: The left-hand input domain.
+			 * 			- \b D2: The right-hand input domain.
+			 * 			- \b D3: The output domain, must be convertible to bool.
+			 * 			- \b operator_type: The internal::operator type to negate.
+			 */
+			template<
+				class Op,
+				enum Backend implementation = config::default_backend
+			>
+			class logical_not {
+				public:
+
+					/** Alias to the left-hand input data type. */
+					typedef typename Op::D1 left_type;
+
+					/** Alias to the right-hand input data type. */
+					typedef typename Op::D2 right_type;
+
+					/** Alias to the output data type. */
+					typedef typename Op::D3 result_type;
+
+					/** Whether this operator has an inplace foldl. */
+					static constexpr bool has_foldl = Op::operator_type::has_foldl;
+
+					/** Whether this operator has an inplace foldr. */
+					static constexpr bool has_foldr = Op::operator_type::has_foldr;
+
+					/**
+					 * Whether this operator is \em mathematically associative; that is,
+					 * associative when assuming equivalent data types for \a IN1, \a IN2,
+					 * and \a OUT, as well as assuming exact arithmetic, no overflows, etc.
+					 */
+					static constexpr bool is_associative = Op::operator_type::is_associative;
```

Comment:
Associativity of the underlying operator is actually not sufficient for the negated operator to be associative -- counterexample:
```
!(!(a or b) or c)=(a or b) and !c=(a and !c) or (b and !c)
while
!(a or !(b or c))=!a and (b or c)=(!a and b) or (!a and c)
```
So e.g., a = true, b=false, and c = false leads to a counter-example.

Reply:
So this means that actually this type trait should be `false`. However, there is something annoying-- consider the type `logical_not< logical_not< OP< T > > >`. The associativity of the doubly-negated `OP` is of course the same as the associativity of `OP` itself.

One potential fix is to introduce an algebraic type trait `is_lnegated<OP>::value` (with lnegated short for logically negated) so that this line can be corrected to:
```
static constexpr const bool is_associative = is_lnegated<OP>::value ? OP::operator_type::is_associative : false;
```

Reply:
See https://github.com/Algebraic-Programming/ALP/pull/224#issuecomment-1637667944